classes/mail/Mailer.inc.php

```diff
@@ -0,0 +1,196 @@
+<?php
+
+/**
+* @file classes/mail/Mailer.inc.php
+*
+ * Copyright (c) 2014-2021 Simon Fraser University
+* Copyright (c) 2000-2021 John Willinsky
+* Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class Mailer
+ * @ingroup mail
+*
+ * @brief Represents interface to manage emails sending and view
+ */
+
+namespace PKP\mail;
+
+use Exception;
+use Illuminate\Contracts\Events\Dispatcher;
+use Illuminate\Contracts\Support\Htmlable;
+use Illuminate\Mail\Mailer as IlluminateMailer;
+use InvalidArgumentException;
+use PKP\i18n\PKPLocale;
+use Swift_Mailer;
+
+class Mailer extends IlluminateMailer
+{
+    const OPEN_TAG = '{';
+    const CLOSING_TAG = '}';
+    const DOLLAR_SIGN_TAG = '$';
+    /**
+     * @var null
+     * Don't bind Laravel View Service, as it's not implemented
+     */
+    protected $views = null;
+
+    /**
+     * @param string $name
+     * @param Swift_Mailer $swift
+     * @param Dispatcher|null $events
+     *
+     * @brief creates new Mailer instance without binding with View
+     */
+    public function __construct(string $name, Swift_Mailer $swift, Dispatcher $events = null)
+    {
+        $this->name = $name;
+        $this->swift = $swift;
+        $this->events = $events;
+    }
+
+    /**
+     * @param string $view
+     * @param array $data variable => value, 'message' is reserved for the Laravel's Swift Message (Illuminate\Mail\Message)
+     * @return string
+     * @brief render email content into HTML string
+     * @throws Exception
+     */
+    protected function renderView($view, $data) : string
+    {
+        if ($view instanceof Htmlable)
+            // return HTML without data compiling
+            return $view->toHtml();
+
+        if (!is_string($view))
+            throw new InvalidArgumentException('View must be instance of ' . Htmlable::class . ' or a string, ' . get_class($view) . ' is given');
+
+        return $this->compileParams($view, $data);
+    }
+
+    /**
+     * @param string $view text or HTML string
+     * @param array $data variables with their values passes ['variable' => value]
+     * @return string compiled string with substitute variables
+     * @throws Exception
+     */
+    public function compileParams(string $view, array $data, string $locale = null) : string
+    {
+        $data = $this->localizeData($data, $locale);
+        unset($data[Mailable::DATA_KEY_MESSAGE]); // Remove pre-set message variable
+
+        $stack = [];
+        $resultString = '';
+        $variable = '';
+        foreach ($template = str_split($view) as $key => $char) {
+            if ($this->openTags($key, $char, $template)) {
+                if (empty($stack)) {
+                    $stack[] = $char;
+                } else {
+                    throw new Exception('Syntax error in email template markdown: opening a new variable tag without closing previous one');
```

Comment:
We may need to handle this more gracefully. Maybe we can perform some validation when an email template is saved. But we'll also need to support cases where a user just makes a typo of some kind.

Reply:
Won't prevent merge.

Reply:
https://github.com/pkp/pkp-lib/issues/7259